src/controllers/ScrollRestorationController.ts

```diff
@@ -0,0 +1,82 @@
+import { Assert } from '../misc/Assert';
+import { ResultListEvents, IDisplayedNewResultsEventArgs } from '../events/ResultListEvents';
+import { QueryStateModel } from '../models/QueryStateModel';
+import { $$ } from '../utils/Dom';
+import { Logger } from '../misc/Logger';
+
+/**
+ * This class is instantiated on the root of {@link SearchInterface}
+ * if the {@link SearchInterface.options.enableScrollRestoration} option is set to true.
+ *
+ * Its job is to wait until the search results are displayed and the page height is long enough to
+ * restore to the last scroll position. It'll wait for up to `restorationTimeOutInMs` milliseconds for the page to reach its height.
+ *
+ * This is useful when back and forth navigation between the search interface and results is needed on the same tab.
+ *
+ */
+
+export class ScrollRestorationController {
+  public static stateAttributeName: string = 'yoffset';
+  /**
+   * Specifies the max amount of time to wait to restore the last scroll position.
+   */
+  public restorationTimeOutInMs: number = 5000;
+
+  private defaultYOffset: number = 0;
+  private tryToScrollIntervalInMs: number = 50;
+  private timeoutHandle = null;
+
+  constructor(public element: HTMLElement, public window: Window, public queryStateModel: QueryStateModel) {
+    Assert.exists(this.queryStateModel);
```

Comment:
Why does this needs to be asserted? 

Usually we assert inputs from developers using the JSUI. If this class is exported and can be instantiated by the frameworks users , then it's helpful for them to assert that the parameters are correct so they will see an error if they are wrong.

On the other side, if it's internal, it should not be necessary as we use Typescript and usually write unit tests that checks those kind of things.

Reply:
> > I prevent scroll restoration if the result list has infinite scrolling enabled, as right now the search interface doesn't preserve any state related to infinite scrolling, but I think it would be cool.
> 
> I'm fairly certain this will cause a problem with most commerce deployment because that's something that a lot of commerce client do.
> 
> It's fine to "not do it right now" in the scope of this PR.
> 
> But if we ship this feature without support for infinite scrolling then it's going to come back to us fairly quickly I'm afraid ;) We have to think about how we can support this.
> 
> > Do you think the class should be a public property on the search interface or leave it internal and expose another component property to set the waiting time out?. I understand that, if possible, you try to avoid adding too many configuration options to components, but maybe in this case makes sense as there is no other need to have this controller public on the SearchInterface.
> 
> The only option I would expose right now is exactly what you have in this PR, an escape hatch to disable/enable the functionality if needed, but having control over timeout and such is too fine grained/not really useful.
> 
> Im thinking maybe we should actually make the default to `true` ... ? 🤔 I'll wait for other team member opinion. It's true that it could cause some unwanted behaviour in some deployment, but on the other hand it's annoying for clients to go and enable this (since you would pretty much always want this enabled, unless there's a bug that's causing issue in your deployment).
> 
> > Browsers have a Back-Forward cache that supposedly caches the html and javascript state when the user leaves the page, but it doesn't work if the unload or beforeunload events are used. And those are used in the codebase, I think that's why the search interface is not preserving the scroll position.
> 
> Can you please confirm this suspicion ? Try and comment out/delete all the unload/beforeunload function you can see in the code, and check what happens. If your suspicion are good, maybe then thinking about how we could remove those function could be a better solution ?
> 
> Have you also tried to test it on a mobile device ? For example, using the `Simulator` app on osx.

Confirmed. After commenting out all the `unload`/`beforeunload` the back-forward cache works beautifully, but just in Firefox and Safari 😕. Not yet implemented on Chrome https://developers.google.com/web/updates/2019/02/back-forward-cache. So, it needs to be done manually until Chrome supports it.

Reply:
Wow the reload time is impressive in the demo/link you added !

But, since we are integrating in existing websites, if we rely only on the non-existence of "unload/beforeunload", it means that if there's any JS in the page that execute such a function (not our code, but any other JS in the website), then it would not work ? So I guess we could not reliably use this, so we would still need to do it manually either way.

But removing the unload/beforeunload part I think could still be part of another story to make page load/navigation time faster for commerce, as well as using more client side caching.

Reply:
Well, yeap, if a client uses unload/beforeunload the back-forward cache it's going to be disabled. It's a pity because it completely preserves the DOM/Javascript state and on back there is no reload/fetch/waiting at all.